drv/fpga-devices/src/ecp5.rs

```diff
@@ -0,0 +1,496 @@
+
+    fn device_enabled(&self) -> Result<bool, FpgaError> {
+        Ok(self.driver.program_n()?)
+    }
+
+    fn set_device_enabled(&self, enabled: bool) -> Result<(), FpgaError> {
+        self.driver.set_program_n(enabled)?;
+        ringbuf_entry!(if enabled {
+            Trace::Enabled
+        } else {
+            Trace::Disabled
+        });
+        Ok(())
+    }
+
+    fn reset_device(&self) -> Result<(), FpgaError> {
+        self.set_device_enabled(false)?;
+        hl::sleep_for(DEVICE_RESET_DURATION);
+        self.set_device_enabled(true)?;
+        Ok(())
+    }
+
+    fn device_state(&self) -> Result<DeviceState, FpgaError> {
+        if !self.driver.program_n()? {
+            Ok(DeviceState::Disabled)
+        } else if self.driver.done()? {
+            Ok(DeviceState::RunningUserDesign)
+        } else if self.driver.init_n()? {
+            Ok(DeviceState::AwaitingBitstream)
+        } else {
+            Ok(DeviceState::Error)
+        }
+    }
+
+    fn device_id(&self) -> Result<u32, FpgaError> {
+        let v = self.read32(Command::ReadId)?;
+        let id = Id::from_u32(v).ok_or(FpgaError::InvalidValue)?;
+        ringbuf_entry!(Trace::ReadId(id));
+        Ok(v)
+    }
+
+    fn start_bitstream_load(&'a self) -> Result<Self::Bitstream, FpgaError> {
+        // Put device in configuration mode if required.
+        if !self.status()?.write_enabled() {
+            self.enable_configuration_mode()?;
+
+            if !self.status()?.write_enabled() {
+                return Err(FpgaError::InvalidState);
+            }
+        }
+
+        // Assert the design reset.
+        self.driver.set_user_design_reset_n(false)?; // Negative asserted.
+
+        // Lock the lower part of the driver in anticipation of writing the
+        // bitstream.
+        let lock = self.lock()?;
+
+        // Use the Impl to write the command and leave the device locked for the
+        // byte stream to follow.
+        self.driver
+            .configuration_write_command(Command::BitstreamBurst)?;
+        ringbuf_entry!(Trace::Command(Command::BitstreamBurst));
+
+        Ok(Ecp5Bitstream {
+            device: self,
+            lock: Some(lock),
+        })
+    }
+}
+
+/// An Ecp5Bitstream type.
+pub struct Ecp5Bitstream<'a, Driver: Ecp5Driver> {
+    device: &'a Ecp5<Driver>,
+    lock: Option<Ecp5Lock<'a, Driver>>,
+}
+
+impl<Driver: Ecp5Driver> Drop for Ecp5Bitstream<'_, Driver> {
+    fn drop(&mut self) {
+        self.lock = None
+    }
+}
+
+impl<Driver: Ecp5Driver> FpgaBitstream for Ecp5Bitstream<'_, Driver>
+where
+    FpgaError: From<<Driver as Ecp5Driver>::Error>,
+{
+    fn continue_load(&mut self, buf: &[u8]) -> Result<(), FpgaError> {
+        self.lock.as_ref().unwrap().0.configuration_write(buf)?;
+        ringbuf_entry!(Trace::WriteBitstreamChunk);
+        Ok(())
+    }
+
+    fn finish_load(&mut self) -> Result<(), FpgaError> {
+        // Release the locked resources in the driver.
+        self.lock = None;
```

Comment:
Should we return an error if `!self.lock.is_none()`?

Reply:
Yes, I added an early return here, because that would mean either a code bug or the user tried to call this out of order (or twice, or both).